### PR TITLE
server/client only environment support

### LINKED ISSUE #57: Don't require client fields
Would be nice if we're just using this for the server if we didn't need to pass in `client` and `clientPrefix`.

```ts
import { createEnv } from "@t3-oss/env-core";
import { z } from "zod";

export const env = createEnv({
    server: {
        DATABASE_URL: z.string().url(),
        OPEN_AI_API_KEY: z.string().min(1),
    },
    client: {}, // this is needed even if it's not used
    clientPrefix: 'client_',  // this is needed even if it's not used
    runtimeEnv: process.env,
});
```

### LINKED ISSUE #63: Next.js apps using `output: "export"` should prevent you passing `server` to `createEnv()`
Hi 👋🏻

Been working on a project where we're using Next.js with `output: "export"` and t3-env and thought it would be great if we could remove the `server` field from `createEnv()` altogether or better yet have it error if you try to pass it in at all.

This is more or less the opposite of what's been mentioned in #57 which could be equally useful.

As t3-env has more support for other frameworks, it's possible they too don't have server environment.

### PR DESCRIPTION
**Pull request by parnavh**
I have opened this pr to add server only and client only functionality which removes the requirement of server/client fields when `serverOnly: boolean` or `clientOnly: boolean` is passed. This aims to addresses #57 and #63. 

I am new to typescript and have come up with an potential implementation, please let me know if this works or any improvements that could be made to my code.

At this moment I have only made changes to the core module and need help porting these changes to nextjs and nuxt modules.

Closes #57
Closes #63

### CODE REVIEW

**Comment by juliusmarminge** [comment]
Thanks for the contribution. Could you

a) resolve conflicts
b) i don't really like an explicit option for this, would be nicer to just let it be possible to omit the client/server entirely, so client only apps would do
```ts
createEnv({
  clientPrefix: "FOO_",
  client: { ... },
});
```
and server-only would be
```ts
createEnv({
  server: { ... },
});
```

**Comment by parnavh** [comment]
Got it, Will do. Also what do you think about @chungweileong94's [view on this](https://github.com/t3-oss/t3-env/issues/57#issuecomment-1556187232) about this package being an overhead for this use case?

**Comment by juliusmarminge** [comment]
> Got it, Will do. Also what do you think about @chungweileong94's https://github.com/t3-oss/t3-env/issues/57#issuecomment-1556187232 about this package being an overhead for this use case?

I do agree with him and have brought it up before too. But that doesn't mean we shouldn't support it ;)

**Comment by juliusmarminge** [comment]
> @juliusmarminge please have a look at the code changes and let me know if something has to changed.
Changes to the tests and docs are pending

Looks promising. Ideally I'd want some tests for it though :)

**Comment by parnavh** [comment]
Can someone from the team have a look, I have made the necessary changes and I feel it is ready

File: packages/core/test/smoke.test.ts [hunk 1]
```diff
@@ -269,3 +269,42 @@ describe("errors when server var is accessed on client", () => {
     );
   });
 });
+
+describe("errors regarding server/client only mode", () => {
+  test("server only", () => {
```
Comment (juliusmarminge) [inline packages/core/test/smoke.test.ts:274]: client only

File: packages/core/test/smoke.test.ts [hunk 2]
```diff
@@ -269,3 +269,42 @@ describe("errors when server var is accessed on client", () => {
...
+    createEnv({
+      clientPrefix: "FOO_",
+      client: {},
+      runtimeEnv: {},
+    });
+  });
+
+  test("client only", () => {
```
Comment (juliusmarminge) [inline packages/core/test/smoke.test.ts:282]: server only

**Comment by juliusmarminge** [comment]
I found a failing test in the core package - it doesn't fail in the Next.js pacakge though.

You have time to investigate?

**Comment by parnavh** [comment]
Oh I thought the tests all passed
I do have time, I'll look into it now, I am new to testing so If I'm unable to get it to work I'll let you know

**Comment by juliusmarminge** [comment]
> Oh I thought the tests all passed

They did, but I added more ;)

**Comment by parnavh** [comment]
Oh alright I'll have a look

**Comment by parnavh** [comment]
I believe this error was introduced in [this change](https://github.com/t3-oss/t3-env/commit/b35b8e87e237ed32d4af3cef3d0d0783fe24f61a#diff-1e2ad102aec04f87784d057de42aaf80a1903c032edf9732e84d7fc7dd58c497) which started checking for prefix in server object. If the `clientPrefix` is empty or in our case undefined it throws an error.

```ts
const env = createEnv({
  server: {
    BAR: z.string(),
  },
  clientPrefix: "",
  client: {},
  runtimeEnv: { BAR: "bar" },
});
```
This code will throw an error since `clientPrefix` is `""` which I don't think should happen. We could deal with empty `clientPrefix` by making change:

```ts
export interface ServerOptions<
  TPrefix extends string,
  TServer extends Record<string, ZodType>
> {
  /**
   * Specify your server-side environment variables schema here. This way you can ensure the app isn't
   * built with invalid env vars.
   */
  server: {
    [TKey in keyof TServer]: TPrefix extends ""
      ? TServer[TKey]
      : TKey extends `${TPrefix}${string}`
      ? ErrorMessage<`${TKey extends `${TPrefix}${string}`
          ? TKey
          : never} should not prefixed with ${TPrefix}.`>
      : TServer[TKey];
  };
}
```

I don't think think this implementation is ideal and can be improved to deal with undefined as well as this only deals with `""`. @juliusmarminge can you help me out with solving this issue please

**Comment by parnavh** [comment]
```ts
server: {
  [TKey in keyof TServer]: TPrefix extends ""
    ? TServer[TKey]
    : TKey extends `${TPrefix}${string}`
    ? ErrorMessage<`${TKey} should not prefixed with ${TPrefix}.`>
    : TServer[TKey];
};
``` 
This code solves the scenario when `clientPrefix: ""` but does not solve it when it is not passed in server only mode.

File: packages/core/test/smoke.test.ts [hunk 3]
```diff
@@ -13,10 +13,10 @@ function ignoreErrors(cb: () => void) {
 
 test("server vars should not be prefixed", () => {
   ignoreErrors(() => {
+    // @ts-expect-error - server should not have FOO_ prefix
```
Comment (juliusmarminge) [inline packages/core/test/smoke.test.ts:16]: this is suboptimal - would love to find a solution where the error is kept on the invalid key as it was, and not the entire object like it is now :/

Comment (juliusmarminge) [reply packages/core/test/smoke.test.ts:16]: actually never mind it didn't work at all :/

File: packages/core/test/smoke.test.ts [hunk 4]
```diff
@@ -269,3 +269,52 @@ describe("errors when server var is accessed on client", () => {
...
+    });
+
+    expectTypeOf(env).toEqualTypeOf<{ BAR: string }>();
+    expect(env).toMatchObject({ BAR: "bar" });
+  });
+
+  test("config with missing client", () => {
+    ignoreErrors(() => {
+      createEnv({
+        // @ts-expect-error - incomplete client config - client not present
```
Comment (juliusmarminge) [inline packages/core/test/smoke.test.ts:302]: this is suboptimal but i couldn't get it working and have perfect errors :/

Comment (juliusmarminge) [reply packages/core/test/smoke.test.ts:302]: let's just roll with it - i doubt this will be an issue really...

**Review by juliusmarminge — APPROVED** [review]

**Pull request merged**